File: mobile_app/lib/models/message.dart
/// Mirrors MessageOut (app/schemas/message.py) — see context/api-contracts.md.
/// Added 2026-09-07 (direct-messaging pass).
class Message {
  Message({
    required this.id,
    required this.patientId,
    required this.senderUserId,
    required this.senderRole,
    required this.senderName,
    required this.content,
    required this.createdAt,
    required this.readAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'] as int,
        patientId: json['patient_id'] as int,
        senderUserId: json['sender_user_id'] as int,
        senderRole: json['sender_role'] as String,
        senderName: json['sender_name'] as String?,
        content: json['content'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
        readAt: json['read_at'] == null ? null : DateTime.parse(json['read_at'] as String),
      );

  final int id;
  final int patientId;
  final int senderUserId;
  final String senderRole; // patient | attendant | doctor
  final String? senderName;
  final String content;
  final DateTime createdAt;
  // Always null for now — no endpoint sets this yet, see
  // app/models/message.py. Kept on the model since the backend returns it.
  final DateTime? readAt;
}

File: Backend/app/schemas/message.py
"""
Schemas for GET/POST /patients/{patient_id}/messages — direct messaging
between a patient (or their attendant) and their assigned doctor. Added
2026-09-07 (direct-messaging pass) — see context/decisions-log.md.

content is bounds-checked the same "generous sanity check, not a real
product limit" spirit as every other free-text field's length bound in
this app (e.g. DoctorNoteCreate has none, but a chat message is
user-facing/repeatedly-sent in a way a clinical note isn't, so a cap here
guards against a runaway client bug rather than a real conversation).
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class MessageCreate(BaseModel):
    content: str = Field(min_length=1, max_length=4000)


class MessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=False)

    id: int
    patient_id: int
    sender_user_id: int
    sender_role: str  # patient | attendant | doctor
    # Computed via a lookup, same "not a plain column" pattern as
    # DoctorNoteOut.doctor_name / PatientOut.assigned_doctor_name.
    sender_name: str | None
    content: str
    created_at: datetime
    # Always null for now — see app/models/message.py's docstring. Exposed
    # here because the task's own schema spec calls for the column, not
    # because anything currently sets it.
    read_at: datetime | None
